### ISSUE TITLE
Raspi

### ISSUE THREAD

**Issue by jpotpo** [deb0ch/CreepyFirmware#1]
has anyone compiled this on raspberry pi ?

**Comment by deb0ch** [comment]
Hi, I haven't tried but it should just work.

I don't see any dependencies that would not be satisfied by the raspberry pi, as the only HW dependency is to have USB ports, one for the usb2dxl servo controller and the other one for the camera.

If anything it _should_ work better on the raspi as I think it has more computing power.

**Comment by jpotpo** [comment]
Thanks for the response I would think so also I am using the usb2ax which I
see you have written code for .
But the creepy firmware will not compile .
I assume it's the main.cpp I should compile ?

**Comment by jpotpo** [comment]
I tried in visual studio won't compile could I have received bad file from
git?

**Comment by deb0ch** [comment]
Hi,

This project is not meant to be compiled in a windows environment but on a Linux.

I seem to remember that Visual Studio uses its own build system, whereas this project uses GNU Make (hence the Makefile). I have never tested or compiled this project on a Windows, so you will likely run into various dependency and build problems.

If you don't have a Linux (dual boot for instance) on your machine I think the least troublesome would be to setup an Ubuntu VM. I think I used 14.04 at the time of writing this project (if 16.04 wasn't out already).

If you do so and it is your first time developing on a Linux, you will see after some time of adaptation that it is a much more developer-friendly environment than Windows (and so is MacOS, and any other system with a decent terminal, shell, standards...).

Although if you really want to stay on Windows and fix the build (you can start trying running the command `make` in a bash-for-windows), you are very welcomed to do so and I would be very glad of receiving your fixes as contributions to this project or the dxlServo library if you ever feel like it 😸

**Comment by jpotpo** [comment]
thanks for the info i have compiled it on the raspi running jessie.
when you addressed the servos is the pan address 1 and tilt address 2 or the opposite?
when i run the program it just seems to run a sine wave ignoring my face so if pan is address 1 it just pans all the way one way stops then pans the other way non stop any pointers here ?

**Comment by deb0ch** [comment]
Oooops I forgot that your initial question was about running it on a Raspi 😅 

The sine wave is when it is in "search mode", searching for a face to track. If it stays there it means that it never detect your face. Maybe there is a problem with either the code, your cam (or your face :trollface: ) ? I already got crappy webcams giving poor results with their low res and very blurry images when in movement.

I don't remember which id is which, they are easily modifiable in the config. The sine wave moving one should be the horizontal move.

Btw, after implementing it I never really liked the sine wave. I think it would have been better off with constant speed and gradual stop. Just never got the motivation to modify it 😁

**Comment by jpotpo** [comment]
ok i have a better camera do you have a way to set the camera id or is that a function of open cv?
do yo recall the code where the search function is so i could disable it for testing?

**Comment by deb0ch** [comment]
IIRC the cam should be identified by its path in `/dev`, which is given as argument to the OpenCV function

**Comment by deb0ch** [comment]
Nvm, it is not that. Apparently it is some number, maybe the number at the end of `/dev/videoX`, X being the number. Or check the docs for the corresponding OpenCV fct because I don't know anymore 😅

**Comment by jpotpo** [comment]
so i tried better camera Logitech 1080 hd detects faces but i think the pi cant keep the frame rate above 2 to 3 fps i noticed in your video yours was running 13 fps .
do you recall what camera you used ?

**Comment by deb0ch** [comment]
Try lowering the resolution inside the program, of the image you're feeding the algorithm. The higher the rez, the heavier the computation will be and you don't need high rez to detect faces.

I think the resolution of a camera is not a characteristic that is relevant for this kind of task, or at least for this kind of task performed using such an algorithm.

For instance, a low exposure time is more interesting so that your pictures taken in movement are not blurry, or the time that your camera takes to focus on an object, or how well it renders images in low light environments, ...

**Comment by jpotpo** [comment]
Ok I'll tweak away at it tonight
Thank you

**Comment by deb0ch** [comment]
FYI, I also used some 1080 hd logitech webcam, and it was not so great. I had very blurry images when the arm was moving due to very high exposure time, needed a lot of light and took like 10 to 30 sec to focus on an object, when something is appearing up close or disappearing.

**Comment by jpotpo** [comment]
Yes same here I'll swap it out

**Comment by deb0ch** [comment]
BTW, there are some issues with the dynamixels that you will likely run into (are you using ax12-a ?) and it might be useful that you know about it.

I have to go now but remind me tomorrow if I haven't written tonight

**Comment by jpotpo** [comment]
Yes the 12a I notice they overheat pretty easy I'll remind you tomorrow
thanks

**Comment by jpotpo** [comment]
Reminder about ax12a

**Comment by deb0ch** [comment]
Yuppp, they overheat pretty easy. Don't trust the built-in alarm, I burnt one without it triggering. Don't trust almost any of the built-in sensors either too much either: torque, heat, voltage, current and so on are either inaccurate or nonexistent (fake data).

When they initialise, they don't always go back to their origin position the direction they should in order to respect their position limits, meaning they might do an extra turn and twist the cables (or beak HW). Have you had that ?

More importantly, have you had any weird jumps during tracking, like the arm going epileptic ?

**Comment by jpotpo** [comment]
I haven't the full spin breaking wires yet .but the epileptic  thing yes
great visual almost choked on soft drink .
 Seems like there isn't much more out there for servos unless you use the
pwm  type
Thanks for the info

**Comment by deb0ch** [comment]
I had this "epileptic" behavior too, I think it is because of the way I was
driving the servos in this software. I don't think the ax12 were designed to be
driven this way, but it is the only solution that I could come up with with the HW
and timeframe available.

I needed to drive them by speed for the position control (the PIDs), but also
needed software position limits to not break the HW or the servos themselves.

In "wheel mode", they take speed inputs but you can't set position limits or
even get a position feedback. So I set them in "position control mode" and
simulate speed orders by sending position setpoints, calculating the delta
position for a delta time to reach that speed.

The problem is (or I _felt_ that was the problem), the ax12-a can't take a new
input before the previous one has been completed. Whenever an input is too big
because of noise, it won't be cancelled / corrected by the next input so your
servo will have a huge jump that will break the tracking.

Not 100% sure it was the problem but it is what I understood of it.

Maybe there is a way of correcting it by software, but the only solution I
can think of was to have an angular position sensor to be able to
set software limits while having the servo in "wheel mode".

**Comment by jpotpo** [comment]
maybe it doesn't like having speed changed and goal position possibly at the same time?
i assume you change speed to catch up to a face but why not a fixed speed that you could change in the config file i assume that is read at boot or would you have to compile each time you change it?
where is the speed changed in the code ?
i will look through it tonight maybe write a fixed speed 
i have a couple more cameras coming i still have to find a perfect fit
